### PR TITLE
Include array bounds checks in SPIR-V output

### PR DESCRIPTION
**Pull request by jimblandy**
This is a recreation of #847, which became orphaned for technical GitHub reasons. I've copied over the unresolved comments here. Still a draft; remaining items:

- [x] Investigate possibly dropped `x` access for `bar.matrix[index].x` in `access.wgsl`.

- [x] Handle structs passed by pointer.

- [ ] Avoid generating many identical `OpArrayLength - 1` expressions.

- [ ] Handle nested structs.

- [ ] Handle arrays sized by pipeline-overrideable constants

- [x] @kvark 's request copied from #847: We need some sort of policy parameter like this:

          enum OutOfBoundsPolicy {
            Restrict, // may be faster than `Avoid`, but we'll need to bench thios
            Avoid,  // portable - matches D3D
            Ignore, // unsafe but low overhead, useful for WebGPU on native (behind a non-portable unsafe feature)
          }

    - [x] Generate proper WGSL zero values for all types.
    - [x] Make stores properly conditional.

### CODE REVIEW

**Comment by jimblandy** [comment]
>     * [x]  Investigate possibly dropped `x` access for `bar.matrix[index].x` in `access.wgsl`.

For `matrix[index].x` we generate three SPIR-V instructions:

    %34 = OpAccessChain  %32  %18 %33 %3
    %35 = OpLoad  %13  %34
    %36 = OpCompositeExtract  %14  %35 0

instead of a single `OpAccessChain`, so I was worried. But the code is correct, if not ideal.

File: src/proc/mod.rs
Comment (Gordon-F) [inline src/proc/mod.rs:146]: How about return `None` here and add warn log with this message?

Comment (jimblandy) [reply src/proc/mod.rs:146]: Hmm, `None` means a dynamically sized array; we take advantage of that in `back/spv/writer.rs`. Maybe this could return `Result<Option<u32>>`?

**Comment by jimblandy** [comment]
Unfortunately, my code for removing redundant length computations was naive, so that had to come out.

But this version implements all three proposed bounds check policies.

**Comment by jimblandy** [comment]
I think the remaining unchecked items are optional, so this is ready for review.

**Comment by jimblandy** [comment]
Support for specialization-constant-sized arrays is still broken in this patch. It won't just fail with an error; it'll try to treat it like a runtime-sized array, and SPIR-V validation will fail. So that is a legit remaining unchecked item.

But in any case, it's ready for review on the rest.

File: src/proc/mod.rs
Comment (jimblandy) [reply src/proc/mod.rs:146]: @kvark What approach would you recommend here?

File: src/back/mod.rs
Comment (kvark) [inline src/back/mod.rs:3]: nit: if it's just one structure, much easier to write it as `serde::Deserialize` there direction instead of adding these lines

File: src/back/mod.rs
Comment (kvark) [inline src/back/mod.rs:48]: This is a tough one to default... and I guess it has to be defaulted, since it's a part of `Options`, where a default is expected.
I think, instead of following the WebGPU requirements here, the default should follow the basic idea of explicit opt-in into more work or features. Since restricting or zeroing is extra code, it's a feature. So the default should be `UndefinedBehavior`.

File: src/back/spv/mod.rs
Comment (kvark) [inline src/back/spv/mod.rs:79]: similarly here, we should default to `UndefinedBehavior`

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:274]: this goes away after rebase, I assume?

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1361]: what does "insn" stand for?

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1438]: we could return this from `write_runtime_array_length` (as a part of the tuple), since it's already getting this type

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1488]: nit: doing `if let ... = ... { ... } else {..}` is a weird substitute for `match`

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1566]: A bit of a problem here: the `write_runtime_array_length` result is actually "length(xx) - 1". So comparing with `<` is incorrect with it.

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1577]: oh, "insns" stands for "instructions"? :)

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1679]: this is quite a bit of logic - it would be great to move it out into a helper

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1685]: a small win here!

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:2146]: similarly, this code should be a candidate for moving out into a helper

File: src/proc/mod.rs
Comment (kvark) [inline src/proc/mod.rs:128]: or a pointer to any of these

File: src/proc/mod.rs
Comment (kvark) [inline src/proc/mod.rs:129]: probably needs a rename to something like `array_access_length` to indicate that we are talking about array-like types here, specifically. Otherwise, it's not clear (judging by the name) why structures are excluded, for example

File: src/proc/mod.rs
Comment (kvark) [reply src/proc/mod.rs:146]: Is this supposed to be called by the both the backends and the frontends? If it's just the backends, moving it to `back/mod.rs` and leaving the unreachable is fine.
If it's both, let's return a result

File: src/valid/function.rs
Comment (kvark) [inline src/valid/function.rs:574]: would it be easier (and still correct) to simply ensure that `TypeFlags::DATA` is there for the base type?
if it's not a structure, the general type validation should see that and complain

File: tests/in/pointer-access.param.ron
Comment (kvark) [inline tests/in/pointer-access.param.ron:1]: any particular reason we are adding a SPIR-V test here (on the input)?
I'd think that bounds checking code is 99% in the backends (as the way we are trying to get it here), so the format of the input test shouldn't matter, in which case we should strongly prefer WGSL on the input.

Can this logic be just merged into the existing `access.wgsl`?

File: tests/out/collatz.spvasm
Comment (kvark) [inline tests/out/collatz.spvasm:4]: We don't want all outputs to show the bounds checks. let's enable it selectively for just a few, if anything, on top of the `access.wgsl`

**Review by kvark — CHANGES_REQUESTED** [review]
Amazing work here! Excited to see this happening 🚀

File: src/proc/mod.rs
Comment (jimblandy) [reply src/proc/mod.rs:146]: It's used in expression validation to do some checks at shader translation time. I'll make it return a `Result`.

File: src/back/mod.rs
Comment (jimblandy) [reply src/back/mod.rs:48]: I'd argue that undefined behavior is something people should have to opt into. It seems to me that the "damn the torpedos, full speed ahead" approach of `UndefinedBehavior` is the feature, and that predictable behavior, following the spec, is the boring status quo, that you get if you're not really paying attention or a newbie.

Comment (jimblandy) [reply src/back/mod.rs:48]: While I do want to push a bit for a safe default here, it's your call. If you think UB by default is a necessity, then I'm happy to make the change.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:274]: It should, yes.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1361]: "Instruction". I think it's a GCC-ism. I'll write it out.

File: src/proc/mod.rs
Comment (jimblandy) [reply src/proc/mod.rs:129]: Makes sense. Renamed to `array_known_length`. I think `known` is an important component of the name.

File: src/back/mod.rs
Comment (kvark) [reply src/back/mod.rs:48]: Yes, I'm convinced Naga should be principled in opting into extra work, here, not safey.
To add more justification, Naga by itself can't generally provide safety, anyway. It relies on the host API, for example to pass the buffer sizes into internal data (like we do in Metal for the `ArrayLength`). The only layer that *can* guarantee safety is `wgpu`, and it's defaulting to safe behaviors. This defaulting includes configuring Naga in a way it needs.

Comment (jimblandy) [reply src/back/mod.rs:48]: Okay, sounds good.

File: src/valid/function.rs
Comment (jimblandy) [reply src/valid/function.rs:574]: You're right, type validation should forbid `ptr<_, array<f32>>`. (It doesn't, but we can fix that.) In that light, the existing check that argument types are sized should handle everything.

Comment (jimblandy) [reply src/valid/function.rs:574]: Actually, we must permit pointers to dynamically sized arrays, since otherwise we have no type for the intermediate `AccessIndex` expression that refers to a dynamically sized array appearing as the last member of a struct.

But making such pointers non-`DATA` (and actually requiring local variables to be `SIZED` `DATA`) prevents the cases we care about preventing.

File: tests/in/pointer-access.param.ron
Comment (jimblandy) [reply tests/in/pointer-access.param.ron:1]: This is SPIR-V only because our WGSL front end doesn't implement the `*` dereferencing operator yet. Once we implement that, it can be replaced with WGSL.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1438]: I added a helper function for producing the unsigned int type.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1488]: The nice thing about `if let else` is that it saves you a level of indentation, compared to match. But I can make this a match if you like.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1566]: I don't think that's correct - `write_runtime_array_length` is the length of the array; it's the `_limit` functions that return one less than the length. I'll change the name of the `_limit` functions to `_max_index`, to make this clearer.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1577]: I've removed this abbreviation.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs]: I took a shot at this, with, I think, mixed results. Let me know what you think of the current version.

File: cli/src/main.rs
Comment (Gordon-F) [inline cli/src/main.rs:84]: Accidentally added feature gate?

Comment (jimblandy) [reply cli/src/main.rs:84]: Oh dear. I'd better look over the rest of this PR more thoroughly. Thank you!

**Comment by jimblandy** [comment]
Okay, fixed the garbage @Gordon-F noticed, and some debugging junk. This, I actually believe is worth reviewing.

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1664]: I'm confused about this. What is the expected type returned here? I'd think it's a boolean, i.e. the function returns something that is a boolean index comparison. But this branch returns an index. How so?

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1754]: best function evar

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1766]: why not return `BoundsCheckResult` from here?

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1784]: similarly, here? I think returning `BoundsCheckResult` would address the other concern I raised, if it's feasible

File: src/proc/mod.rs
Comment (kvark) [inline src/proc/mod.rs:121]: can we move all the indexable logic into `src/proc/index.rs` and re-export here?

File: tests/out/pointer-access.info.ron
Comment (kvark) [inline tests/out/pointer-access.info.ron:2]: do we need these RON snapshots for pointer-access? They are fairly verbose.
I think we should omit them unless there is a strong reason to include them

**Review by kvark — COMMENTED** [review]
Great! The code is shaping up just beautifully ❤️ 
Just a couple organizational comments. Overall looking quite landable.

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs:1508]: To avoid leaking a bunch of `writer`-local types, this needed to be a sub-module of `writer`: `pub(in ...)` can only restrict visibility to parents of the module containing the definition. So it's `src/back/spv/writer/index.rs`, which is a bit deep, but not the end of the world.

Comment (jimblandy) [reply src/back/spv/writer.rs:1508]: If making `index` a child module of `writer` is too much nesting, then there's the alternative of adding a bunch of non-`pub` `use` declarations to `src/back/spv/mod.rs`. I didn't explore that possibility.

File: src/proc/mod.rs
Comment (jimblandy) [reply src/proc/mod.rs:121]: Sure.

File: tests/out/pointer-access.info.ron
Comment (jimblandy) [reply tests/out/pointer-access.info.ron:2]: Okay, sounds good.

**Review by jimblandy — COMMENTED** [review]
Okay, I think I've addressed all the comments. New push in a sec.

**Comment by jimblandy** [comment]
Okay, hmm, I wasn't expecting GitHub to treat my replies as a separate review. I'll be sure not to do that again.

Upshot was, all comments addressed as expected.

**Comment by jimblandy** [comment]
Lovely. I can't reproduce those clippy complaints locally. Need to walk dogs, will fix once I get back.

File: src/back/spv/writer.rs
Comment (kvark) [reply src/back/spv/writer.rs]: yeah, I don't think we need this level of nesting in either parts of Naga.
Writer isn't meant to be this huge sub-module of SPV-out. It's just accidental that it has all the logic today. Coincidentally, "writer-local" is pretty much "spv-in" namespace, since "writer" dominates it. So `pub(super)` would work ok here.

File: src/back/spv/writer.rs
Comment (kvark) [inline src/back/spv/writer.rs:1]: can we make it just `back/spv/index.rs` instead of `back/spv/writer/index.rs`?

**Comment by jimblandy** [comment]
(just trying to placate clippy and accommodate changes to error behavior on master; some review comments not yet addressed)

File: src/back/spv/writer.rs
Comment (jimblandy) [reply src/back/spv/writer.rs]: I've implemented "approach C", and it was fine.

**Comment by jimblandy** [comment]
Okay - that's all green, and with the indexing code in `src/back/spv/index.rs`. I think this is ready.

**Comment by jimblandy** [comment]
For ease of review, I separated out moving `Writer` from `writer.rs` to `mod.rs` into its own commit, which (hopefully) includes no significant code changes. So if you look at the individual commits, you'll get a more meaningful view of the actual work.

**Comment by jimblandy** [comment]
(rebase onto latest master)

File: src/back/spv/index.rs
Comment (kvark) [inline src/back/spv/index.rs:4]: nit: we generally use one `use` for each thing (like `super`)

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
Waiting for #990 and the release happen first

**Pull request merged**
